aws-lc-rs/src/rsa/encryption.rs

```diff
@@ -0,0 +1,474 @@
+type Mgf1HashFn = unsafe extern "C" fn() -> *const EVP_MD;
+
+/// An RSA-OAEP algorithm.
+pub struct OaepAlgorithm {
+    id: EncryptionAlgorithmId,
+    oaep_hash_fn: OaepHashFn,
+    mgf1_hash_fn: Mgf1HashFn,
+}
+
+impl OaepAlgorithm {
+    /// Returns the `EncryptionAlgorithmId`.
+    #[must_use]
+    pub fn id(&self) -> EncryptionAlgorithmId {
+        self.id
+    }
+
+    #[inline]
+    fn oaep_hash_fn(&self) -> OaepHashFn {
+        self.oaep_hash_fn
+    }
+
+    #[inline]
+    fn mgf1_hash_fn(&self) -> Mgf1HashFn {
+        self.mgf1_hash_fn
+    }
+}
+
+impl Debug for OaepAlgorithm {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        Debug::fmt(&self.id, f)
+    }
+}
+
+/// An RSA private key used for decrypting ciphertext encrypted by a [`PublicEncryptingKey`].
+pub struct PrivateDecryptingKey(LcPtr<EVP_PKEY>);
+
+impl PrivateDecryptingKey {
+    fn new(evp_pkey: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
+        Self::validate_key(&evp_pkey)?;
+        Ok(Self(evp_pkey))
+    }
+
+    fn validate_key(key: &LcPtr<EVP_PKEY>) -> Result<(), Unspecified> {
+        if !is_rsa_key(key) {
+            return Err(Unspecified);
+        };
+        match key_size_bits(key) {
+            2048..=8192 => Ok(()),
+            _ => Err(Unspecified),
+        }
+    }
+
+    /// Generate a new RSA private key for use with asymmetrical encryption.
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs during the generation of the RSA keypair.
+    pub fn generate(size: KeySize) -> Result<Self, Unspecified> {
+        let key = generate_rsa_key(size.bit_len(), false)?;
+        Self::new(key)
+    }
+
+    /// Generate a RSA `KeyPair` of the specified key-strength.
+    ///
+    /// Supports the following key sizes:
+    /// * `EncryptionKeySize::Rsa2048`
+    /// * `EncryptionKeySize::Rsa3072`
+    /// * `EncryptionKeySize::Rsa4096`
+    ///
+    /// # Errors
+    /// * `Unspecified`: Any key generation failure.
+    #[cfg(feature = "fips")]
+    pub fn generate_fips(size: KeySize) -> Result<Self, Unspecified> {
+        let key = generate_rsa_key(size.bit_len(), true)?;
+        Self::new(key)
+    }
+
+    /// Construct a `PrivateDecryptingKey` from the provided PKCS#8 (v1) document.
+    ///
+    /// Supports RSA key sizes between 2048 and 8192 (inclusive).
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs during deserialization of this key from PKCS#8.
+    pub fn from_pkcs8(pkcs8: &[u8]) -> Result<Self, Unspecified> {
+        let evp_pkey = LcPtr::try_from(pkcs8)?;
+        Self::new(evp_pkey)
+    }
+
+    /// Returns a boolean indicator if this RSA key is an approved FIPS 140-3 key.
+    #[cfg(feature = "fips")]
+    #[must_use]
+    pub fn is_valid_fips_key(&self) -> bool {
+        super::key::is_valid_fips_key(&self.0)
+    }
+
+    /// Returns the RSA key size in bytes.
+    #[must_use]
+    pub fn key_size(&self) -> usize {
```

Comment:
Since RSA "key size" is almost always talked about in "bits", calling this `key_size` is ambiguous. I think we should provide both `key_size_bits` and `key_size_bytes`.